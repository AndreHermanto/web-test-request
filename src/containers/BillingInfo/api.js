/**
 * Returns the pricing based on query strings - panelId, payer and familyExtras
 * @return {Object[]} json Currently returns an array of test panel object - including id and label
 */
export function getPricing(panelId, type, payer, familyExtras) {
  return fetch(`${process.env.REACT_APP_BASE_URL}/pricing?panelId=${panelId}&geneListType=${type}&payer=${payer}&familyExtras=${familyExtras}`)
    .then((response) => response.json())
    .then((json) => json.data);
}