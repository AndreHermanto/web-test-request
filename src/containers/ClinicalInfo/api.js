/**
 * Upload an attachment
 * @param {Object} data Test request data to be submitted.
 */
export function upload(file) {
  var formData = new FormData();
  formData.append('file', file);
  return fetch(`${process.env.REACT_APP_BASE_URL}/attachments`, {
    method: 'POST',
    body: formData
  }).then((response) => response.json() );
}
