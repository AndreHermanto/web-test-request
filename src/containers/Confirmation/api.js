/**
 * Returns the test request based on id
 * @return {Object[]} json Currently returns an array of test panel object - including id and label
 */
export function getTestRequestById(id) {
  return fetch(`${process.env.REACT_APP_BASE_URL}/test-requests/${id}`)
    .then((response) => response.json())
    .then((json) => json.data);
}
