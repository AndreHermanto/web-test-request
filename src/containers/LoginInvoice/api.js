/**
 * Returns an invoice. 
 */
export function getInvoice(user) {
  return fetch(`${process.env.REACT_APP_BASE_URL}/test-requests?id=${user.id}&mobilePhone=${user.phone}&lastName=${user.lastName}`)
    .then((response) => response.json())
    .then((json) => json.data);
}
