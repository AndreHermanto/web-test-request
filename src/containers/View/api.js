/**
 * Returns the test request based on id
 * @return {Object[]} json Currently returns an array of test panel object - including id and label
 */
export function getTestRequestById(id, username, password) {
  return fetch(`${process.env.REACT_APP_BASE_URL}/test-requests/${id}`, {
      headers: {
        Authorization: 'Basic ' + btoa(username + ':' + password)  
      }
    })
    .then((response) => response.json())
    .then((json) => json.data);
}

/**
 * Returns the test request based on firstname, lastname and dob
 * @return {Object[]} json Currently returns an array of test panel object - including id and label
 */
export function getTestRequestByPatientInfo(firstName, lastName, dob, username, password) {
  return fetch(`${process.env.REACT_APP_BASE_URL}/test-requests?firstName=${firstName}&lastName=${lastName}&dob=${dob}`, {
      headers: {
        Authorization: 'Basic ' + btoa(username + ':' + password)  
      }
    })
    .then((response) => response.json())
    .then((json) => json.data);
}