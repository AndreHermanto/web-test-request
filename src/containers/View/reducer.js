import validator from './../../components/validator';

export function initData() {
  var state = {
    form: {
      searchBy:'Test Request Id',
      testRequestId:'',
      lastName: '',
      firstName: '',
      dob: ''
    },
    testRequestData: {},
    testRequestList: [],
    validationRule: {
      testRequestId: 'required',
      lastName: 'required',
      firstName: 'required',
      dob: 'requiredDate'
    },
    validated: false,
    formId: 'ViewModule'
  };
  
  // This validates the data in the initial state.
  state.validation = {};
  for (var field in state.validationRule) {
    if(field) {
      state.validation[field] = validator(state.form[field], state.validationRule[field]);
    }
  }
  
  return Object.assign({}, state);
}

/**
 * This sets the form data upon onChange.
 * @param {Object} state Targeted state to be changed.
 * @param {Object} target Target object captured from UI event change.
 */
export function setFormData(state, target) {
  var value = target.value;
  
  var formStateChild = Object.assign({}, state.form, {
    [target.name]: value
  });
  
  var validationStateChild = Object.assign({}, state.validation, {
    [target.name]: validator(value, state.validationRule[target.name])
  });

  return Object.assign({}, state, {
    form: formStateChild,
    validation: validationStateChild,
    validated: false
  });
}

/**
 * This sets testRequestList state.
 * @param {Object} state Targeted state to be changed.
 * @param {Object[]} list A list of test request data pulled from the backend
 */
export function setTestRequestList(state, list) {
  return Object.assign({}, state, {
    testRequestList: list
  });
}

/**
 * Set "validated" state to true - identifying the confirm button is clicked and validation processed.
 * @param {Object} state Targeted state to be changed.
 */
export function validatedToTrue(state) {
  return Object.assign({}, state, {
    validated: true,
    testRequestList: []
  });
}