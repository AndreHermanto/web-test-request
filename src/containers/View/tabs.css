#testRequestSearch .nav-stacked {
  padding: 0 !important;
}

#testRequestSearch .nav-stacked li {
  border-top: 1px solid #ccc;
  margin-top: 0 !important;
}

#testRequestSearch .nav-stacked li:first-child {
  border-top: none;
}

#testRequestSearch .nav-stacked .active a {
  background: #00a6b6;
  border: none !important;
  border-radius: 0 !important;
}

#testRequestSearch .nav-stacked li:hover a {
  border: none !important;
  border-radius: 0 !important;
}
