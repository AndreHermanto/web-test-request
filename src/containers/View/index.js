import React, { Component } from 'react';
import { 
  Tab,
  Nav,
  NavItem
} from 'react-bootstrap';
import styled from 'styled-components';
import {
  initData,
  setFormData,
  setTestRequestList,
  validatedToTrue
} from './reducer';
import {
  getTestRequestById,
  getTestRequestByPatientInfo
} from './api';
import OrderTestModule from './../Summary/components/OrderTestModule';
import PatientDetailsModule  from './../Summary/components/PatientDetailsModule';
import FamilyMemberModule from './../Summary/components/FamilyMemberModule';
import ClinicianDetailsModule  from './../Summary/components/ClinicianDetailsModule';
import BillingInfoModule from './../Summary/components/BillingInfoModule';
import { 
  FormButton,
  PageHeading
} from './../../components/SharedStyle';
import Input from './../../components/Input';
import RadioSet from './../../components/RadioSet';
import DatePicker from './../../components/DatePicker';
import './tabs.css';

const SearchArea = styled.div`
  width: 35%;
  float: left;
  overflow: hidden;
  padding-right: 24px;
  padding-left: 4px;
  padding-top: 0px;
  margin-top: 0px;
  border-right: thin solid #e7e7e7;
`;

const TestRequestView = styled.div`
  width: 65%
  padding-left: 6%;
  float: left;
  min-height: 450px;
`;

const ListContainer = styled(Nav)`
  padding: 4px !important;
  border-radius: 0px !important;
  background-color: #fff;
  border: 1px solid;
  margin: 6px 0 6px 0;
  cursor: pointer;
  color:${props => props.selected ? '#00a6b6' : '#000'};
  border-color: ${props => props.selected ? '#00a6b6' : '#ccc'};
  &:hover {
    border-color: #00a6b6;
    color: #00a6b6 !important;
  }

  li {
    border-color: #00a6b6;
    color: #00a6b6 !important;
  }
`;

/**
* Confirmation - UI for ordering type of tests, selecting disorder and related genes for testing.
*/
class View extends Component {
  constructor(props) {
    super(props);
    this.handleChange = this.handleChange.bind(this);
    this.handleSearchById = this.handleSearchById.bind(this);
    this.handleSearchByPatientDetails = this.handleSearchByPatientDetails.bind(this);
    this.state = initData();
  }
  
  handleChange(event) {
    this.setState(setFormData(this.state, event.target),
    this.setState(setTestRequestList(this.state, [])));
  }
  
  handleSearchById() {
    this.setState(validatedToTrue(this.state));
    
    if(
      this.state.validation.testRequestId.status !== 'error'
    ) {
      return getTestRequestById(this.state.form.testRequestId)
        .then((data) => {
          var array = [];
          if(data.id) array.push(data);
          this.setState(setTestRequestList(this.state, array));
        });
    }
    
    return false;
  }
  
  handleSearchByPatientDetails() {
    this.setState(validatedToTrue(this.state));
    
    if(
      this.state.validation.firstName.status !== 'error' &&
      this.state.validation.lastName.status !== 'error' &&
      this.state.validation.dob.status !== 'error'
    ) {                              
      return getTestRequestByPatientInfo(
        this.state.form.firstName,
        this.state.form.lastName,
        this.state.form.dob
      )
        .then((data) => {
          this.setState(setTestRequestList(this.state, data));
        });  
    }
                                  
    return false;
  }
  
  // This renders the validation result after confirm button is clicked.
  validate() {
    return this.state.validated && this.state.validation;
  }
  
  render() {
    return (
      <Tab.Container id="testRequestSearch">
        <div>
          <SearchArea>
            <PageHeading>Search Test Request</PageHeading>

            <RadioSet 
              field="searchBy"
              label="Search by"
              options={['Test Request Id', 'Patient Details']}
              onChange={this.handleChange}
              formState={this.state.form}
              inline
            />

            <hr />

            {this.state.form.searchBy === 'Test Request Id' &&
              <div>
                <Input
                  field="testRequestId"
                  label="Test Request Id"
                  onChange={this.handleChange}
                  onValidate={this.validate()}
                  formState={this.state.form}
                />

                <FormButton  
                  type="submit" 
                  onClick={this.handleSearchById}
                >
                  Search
                </FormButton> 
              </div>
            }

            {this.state.form.searchBy === 'Patient Details' &&
              <div>
                <Input
                  field="firstName"
                  label="Given Name"
                  onChange={this.handleChange}
                  onValidate={this.validate()}
                  formState={this.state.form}
                />

                <Input
                  field="lastName"
                  label="Surname"
                  onChange={this.handleChange}
                  onValidate={this.validate()}
                  formState={this.state.form}
                />

                <DatePicker
                  field="dob"
                  label="Date of Birth"
                  onChange={this.handleChange}
                  onValidate={this.validate()}
                  formState={this.state.form}
                />

                <FormButton  
                  type="submit" 
                  onClick={this.handleSearchByPatientDetails}
                >
                  Search
                </FormButton> 
              </div>
            }

            {(this.state.testRequestList && this.state.validated) &&
              <div>
                <br /><hr />
                <PageHeading>Found Test Request</PageHeading>
                {
                  (this.state.testRequestList.length > 0) &&
                  <ListContainer bsStyle="pills" stacked>
                  {
                    this.state.testRequestList.map((testRequest, $index) => {
                      return <NavItem eventKey={'tr'+$index} key={'tr'+$index}>
                        <div>ID: {testRequest.id}</div>
                        <div>Submitted time: {testRequest.createdDateTime}</div>
                      </NavItem>
                    })
                  }
                  </ListContainer>
                }
                
                {
                  (this.state.testRequestList.length === 0) &&
                  <p>
                    No test request is found.
                  </p>
                }
              </div>
            }
          </SearchArea>

          <TestRequestView>
            {(this.state.testRequestList && this.state.testRequestList.length === 0) &&
              <p><br /><br />Please use the left hand panel to search for test request. </p>
            }
            
            {(this.state.testRequestList && this.state.testRequestList.length > 0) &&
              <Tab.Content>
                {
                  this.state.testRequestList.map((testRequest, $index) => {
                    return <Tab.Pane eventKey={'tr'+$index} key={'tr'+$index} >
                      <OrderTestModule orderTestModule={testRequest.orderTestModule}/>
                      <PatientDetailsModule patientDetails={testRequest.patientDetailsModule} clinicalInfo={testRequest.clinicalInfoModule}/>
                      {
                        testRequest.familyMembersModule.familyMembers.length > 0 &&
                        testRequest.familyMembersModule.familyMembers.map((member, i) => 
                        {
                          return <FamilyMemberModule familyMemberDetails={member.familyMemberDetails} familyMemberClinicalInfo={member.familyMemberClinicalInfo} key={i}/>
                        })
                      }
                      <ClinicianDetailsModule clinicianDetailsModule={testRequest.clinicianDetailsModule}/>
                      <BillingInfoModule billingInfoModule={testRequest.billingInfoModule} clinicianDetailsModule={testRequest.clinicianDetailsModule}/>
                    </Tab.Pane>
                  })
                }
              </Tab.Content>
            }
          </TestRequestView>
        </div>
      </Tab.Container>
    );
  }
}

export default View;