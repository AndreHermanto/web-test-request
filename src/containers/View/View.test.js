import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import TestUtils from 'react-addons-test-utils';
import View from './index';

jest.unmock('./api.js');
import * as api from './api';

var mockData = [
  {
  "orderTestModule": {
        "test": {
            "id": "58c74b7c1020a1eb58e9dfac",
            "label": "Dilated Cardiomyopathy (DCM), Core Panel",
            "geneLists": []
        }
    },
    "patientDetailsModule": {
        "lastName": "Agudo",
        "firstName": "Joshua",
        "dob": "2020-02-02",
        "medicalRecordNo": "",
        "gender": "Male",
        "genderOther": "",
        "ethnicity": "Filipino",
        "deceased": false,
        "sampleSource": "",
        "consent": true
    },
    "clinicalInfoModule": {
        "clinicalInfo": "Here is some clinical information",
        "affected": false,
        "consanguinityInfo": "asjhdasbdlkahsdjb",
        "familyHistory": "kasdfbljkbfalhfasvfkjabskldfbkjsdfbkdashfvdjhfhb;kfjv;hfv;fjbafhb;akfd",
        "relevantInvestigation":"asdkfhgasdlfhabsdlfjhalsfdjh",
        "consanguinity": false
    },
    "familyMembersModule": {
        "familyMembers": [
            {
                "familyMemberDetails": {
                    "lastName": "Blah",
                    "firstName": "Mother",
                    "dob": "2020-02-02",
                    "medicalRecordNo": "adsa",
                    "gender": "Female",
                    "genderOther": "",
                    "ethnicity": "",
                    "deceased": false,
                    "sampleSource": "",
                    "consent": true
                },
                "familyMemberClinicalInfo": {
                    "clinicalInfo": "adssda",
                    "affected": true,
                    "consanguinityInfo": "asjhdasbdlkahsdjb",
                    "familyHistory": "kasdfbljkbfalhfasvfkjabskldfbkjsdfbkdashfvdjhfhb;kfjv;hfv;fjbafhb;akfd",
                    "relevantInvestigation":"asdkfhgasdlfhabsdlfjhalsfdjh",
                    "consanguinity": false
                }
            }
        ]
    },
    "clinicianDetailsModule": {
        "providerNumber": "123123",
        "medicalSpecialty": "IBD",
        "firstName": "Boaty",
        "lastName": "McBoatFace",
        "organisation": "Garvan Institute",
        "phone": "12390098123",
        "email": "slkjs@alkjs.com",
        "fax": "",
        "copy": false,
        "copyToHCP": [
                {
                    "firstName": "Hello",
                    "lastName": "World",
                    "organisation": "Some organisation",
                    "email": "hello@world.com"
                }
        ],
    },
    "billingInfoModule": {
        "billOption": "Private",
        "payer": "Joshua Agudo",
        "phone": "1231232131",
        "firstName": "Joshua",
        "lastName": "Agudo",
        "payerEmail": "joshua.agudo@genome.one",
        "consent": true
    },
    "signature": false,
    "id": 1
  }
];

describe('View test', () => {
  test('View page renders without crashing', () => {
    const page = renderer.create(React.createElement(View)).toJSON();
    expect(page).toMatchSnapshot();
  });
  
  test('handleChange works', () => {
    const page = TestUtils.renderIntoDocument(React.createElement(View));
    const selection = { "target": { 
      "value": 1,
      "name": "testRequestId"
    }};
    page.handleChange(selection);
    expect(page.state.form.testRequestId).toEqual(1);
    expect(page.state.testRequestList.length).toEqual(0); 
  });
  
  test('handleSearchById search for one test request',  () => {
    api.getTestRequestById = jest.fn(() => {
      return new Promise( resolve => 
        resolve(mockData[0])
      );
    })
    const page = TestUtils.renderIntoDocument(React.createElement(View));
    const selection1 = { "target": { 
      "value": "admin",
      "name": "username"
    }};
    const selection2 = { "target": { 
      "value": "secret",
      "name": "password"
    }};
    const selection3 = { "target": { 
      "value": 1,
      "name": "testRequestId"
    }};
    page.handleChange(selection1);
    page.handleChange(selection2);
    page.handleChange(selection3);

    page.handleSearchById()
    expect(api.getTestRequestById).toBeCalled();
  });
  
  test('handleSearchById search fails occur when no input is done',  () => {
    const page = TestUtils.renderIntoDocument(React.createElement(View));
    api.getTestRequestById = jest.fn(() => {
      return new Promise( resolve => 
        resolve(mockData[0])
      );
    })
    const selection1 = { "target": { 
      "value": "admin",
      "name": "username"
    }};
    const selection2 = { "target": { 
      "value": "secret",
      "name": "password"
    }};
    page.handleChange(selection1);
    page.handleChange(selection2);
    page.handleSearchById();
    expect(api.getTestRequestById).not.toBeCalled();
  });
  
  test('handleSearchByPatientDetails search for one test request', () => {
    api.getTestRequestByPatientInfo = jest.fn(() => {
      return new Promise( resolve => 
        resolve(mockData)
      );
    })
    const page = TestUtils.renderIntoDocument(React.createElement(View));
    const selection1 = { "target": { 
      "value": "admin",
      "name": "username"
    }};
    const selection2 = { "target": { 
      "value": "secret",
      "name": "password"
    }};
    const selection3 = { "target": { 
      "value": "Joshua",
      "name": "firstName"
    }};
    const selection4 = { "target": { 
      "value": "Agudo",
      "name": "lastName"
    }};
    const selection5 = { "target": { 
      "value": "2020-02-02",
      "name": "dob"
    }};
    page.handleChange(selection1);
    page.handleChange(selection2);
    page.handleChange(selection3);
    page.handleChange(selection4);
    page.handleChange(selection5);
    page.handleSearchByPatientDetails();
    expect(api.getTestRequestByPatientInfo).toBeCalled();
  });
  
  test('handleSearchByPatientDetails search fails occur when no input is done', () => {
    api.getTestRequestByPatientInfo = jest.fn(() => {
      return new Promise( resolve => 
        resolve(mockData)
      );
    })
    const page = TestUtils.renderIntoDocument(React.createElement(View));
    page.handleSearchByPatientDetails();
    expect(api.getTestRequestByPatientInfo).not.toBeCalled();
  });
})