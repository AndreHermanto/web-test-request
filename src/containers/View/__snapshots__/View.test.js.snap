exports[`View test View page renders without crashing 1`] = `
<div
  id="testRequestSearch">
  <div
    className="dQyqsI">
    <h3
      className="hWxLNB">
      Search Test Request
    </h3>
    <div
      className="form-group"
      onChange={[Function]}>
      <label
        className="control-label"
        htmlFor={undefined}>
        Search by
      </label>
      <br />
      <label
        className="radio-inline"
        style={
          Object {
            "marginRight": 8,
          }
        }>
        <input
          defaultChecked={true}
          disabled={false}
          name="searchBy"
          type="radio"
          value="Test Request Id" />
        Test Request Id
      </label>
      <label
        className="radio-inline"
        style={
          Object {
            "marginRight": 8,
          }
        }>
        <input
          defaultChecked={false}
          disabled={false}
          name="searchBy"
          type="radio"
          value="Patient Details" />
        Patient Details
      </label>
      <span
        className="SfhHK help-block" />
    </div>
    <hr />
    <div>
      <div
        className="form-group has-feedback">
        <label
          className="control-label"
          htmlFor={undefined}>
          Test Request Id
        </label>
        <input
          className="fkPJON form-control"
          disabled={undefined}
          id={undefined}
          name="testRequestId"
          onChange={[Function]}
          type="text"
          value="" />
        <span
          className="SfhHK help-block" />
      </div>
      <button
        className="bBdbjl"
        onClick={[Function]}
        type="submit">
        Search
      </button>
    </div>
  </div>
  <div
    className="kLsVgF">
    <p>
      <br />
      <br />
      Please use the left hand panel to search for test request. 
    </p>
  </div>
</div>
`;
